gobblin-modules/gobblin-parquet/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,527 @@
+
+    protected JsonElementConverter(JsonSchema schema, WorkUnitState state) {
+      this.jsonSchema = schema;
+      this.state = state;
+    }
+
+    /**
+     * Convert value to a parquet type and perform null check.
+     * @param value
+     * @return Parquet safe type
+     */
+    public Object convert(JsonElement value) {
+      if (value.isJsonNull()) {
+        if (this.jsonSchema.isNullable()) {
+          return null;
+        }
+        throw new RuntimeException(
+            "Field: " + this.jsonSchema.getColumnName() + " is not nullable and contains a null value");
+      }
+      return convertField(value);
+    }
+
+    /**
+     * Returns a {@link Type} parquet schema
+     * @return
+     */
+    abstract public Type schema();
+
+    /**
+     * Convert JsonElement to Parquet type
+     * @param value
+     * @return
+     */
+    abstract Object convertField(JsonElement value);
+  }
+
+  /**
+   * Converts a {@link JsonSchema} to a {@link PrimitiveType}
+   */
+  public static abstract class PrimitiveConverter extends JsonElementConverter {
+    protected final boolean repeated;
+    private PrimitiveTypeName outputType;
+    protected Type schema;
+
+    /**
+     * @param jsonSchema
+     * @param repeated
+     * @param outputType
+     * @param state
+     */
+    public PrimitiveConverter(JsonSchema jsonSchema, boolean repeated, PrimitiveTypeName outputType,
+        WorkUnitState state) {
+      super(jsonSchema, state);
+      this.repeated = repeated;
+      this.outputType = outputType;
+      this.schema = buildSchema();
+    }
+
+    private Type buildSchema() {
+      return new PrimitiveType(this.repeated ? REPEATED : this.jsonSchema.optionalOrRequired(), this.outputType,
+          this.jsonSchema.getColumnName());
+    }
+  }
+
+  /**
+   * Converts {@link JsonSchema} having collection of elements of {@link InputType} into a {@link GroupType}.
+   */
+  public static abstract class CollectionConverter extends JsonElementConverter {
+    protected InputType elementType;
+    protected JsonElementConverter elementConverter;
+    protected Type schema;
+
+    public CollectionConverter(JsonSchema collectionSchema, InputType elementType, WorkUnitState state,
+        boolean repeated) {
+      super(collectionSchema, state);
+      this.elementType = elementType;
+      this.elementConverter = getConverter(getElementSchema(), state, repeated);
+    }
+
+    /**
+     * Prepare a {@link JsonSchema} for the elements in a collection.
+     * @return
+     */
+    abstract JsonSchema getElementSchema();
+  }
+
+  public static class IntConverter extends PrimitiveConverter {
+
+    public IntConverter(JsonSchema schema, boolean repeated, WorkUnitState state) {
+      super(schema, repeated, INT32, state);
+    }
+
+    @Override
+    IntegerValue convertField(JsonElement value) {
+      return new IntegerValue(value.getAsInt());
+    }
+
+    @Override
+    public Type schema() {
```

Comment:
Can be moved to `PrimitiveConverter`. The comment applied to the other primitive converters